web/pandas/pdeps/00xx-string-dtype.md

```diff
@@ -0,0 +1,253 @@
+
+## Alternatives
+
+### Why not delay introducing a default string dtype?
+
+To avoid introducing a new string dtype while other discussions and changes are
+in flux (eventually making pyarrow a required dependency? adopting `pd.NA` as
+the default missing value sentinel? using the new NumPy 2.0 capabilities?), we
+could also delay introducing a default string dtype until there is more clarity
+in those other discussions.
+
+However:
+
+1. Delaying has a cost: it further postpones introducing a dedicated string
+   dtype that has massive benefits for our users, both in usability as (for the
+   significant part of the user base that has PyArrow installed) in performance.
+2. In case we eventually transition to use `pd.NA` as the default missing value
+   sentinel, we will need a migration path for _all_ our data types, and thus
+   the challenges around this will not be unique to the string dtype and
+   therefore not a reason to delay this.
+
+### Why not use the existing StringDtype with `pd.NA`?
+
+Wouldn't adding even more variants of the string dtype will make things only more
+confusing? Indeed, this proposal unfortunately introduces more variants of the
+string dtype. However, the reason for this is to ensure the actual default user
+experience is _less_ confusing, and the new string dtype fits better with the
+other default data types.
+
+If the new default string data type would use `pd.NA`, then after some
+operations, a user can easily end up with a DataFrame that mixes columns using
+`NaN` semantics and columns using `NA` semantics (and thus a DataFrame that
+could have columns with two different int64, two different float64, two different
+bool, etc dtypes). This would lead to a very confusing default experience.
+
+With the proposed new variant of the StringDtype, this will ensure that for the
+_default_ experience, a user will only see only 1 kind of integer dtype, only
+kind of 1 bool dtype, etc. For now, a user should only get columns with an
+`ArrowDtype` and/or using `pd.NA` when explicitly opting into this.
+
+## Backward compatibility
+
+The most visible backwards incompatible change will be that columns with string
+data will no longer have an `object` dtype. Therefore, code that assumes
+`object` dtype (such as `ser.dtype == object`) will need to be updated.
+
+To allow testing your code in advance, the
+`pd.options.future.infer_string = True` option is available.
+
+Otherwise, the actual string-specific functionality (such as the `.str` accessor
+methods) should all keep working as is. By preserving the current missing value
+semantics, this proposal is also backwards compatible on this aspect.
+
+One other backwards incompatible change is present for early adopters of the
+existing `StringDtype`. In pandas 3.0, calling `pd.StringDtype()` will start
+returning the new default string dtype, while up to now this returned the
+experimental string dtype using `pd.NA` introduced in pandas 1.0. Those users
```

Comment:
Historically you would get this by using `dtype="string"` too right? I'm a little wary that we are underestimating the scope of how breaking this could be; I didn't even realize we considered that dtype experimental all this time

Reply:
A deprecation warning about what exactly?

Reply:
> at the expense of increasing the scope of breaking changes for the smaller subset of users that was already using `dtype="string"`

This is where I am a little uncomfortable - I don't know how to measure the size of that, but I am wary of assuming it is not a signifcant number of users. The fact that "string" returns NA as a missing value is a documented difference in our code base:

https://pandas.pydata.org/docs/dev/user_guide/text.html#behavior-differences

And its usage has been promoted for quite some time:

https://stackoverflow.com/a/60553529/621736
https://towardsdatascience.com/why-we-need-to-use-pandas-new-string-dtype-instead-of-object-for-textual-data-6fd419842e24
https://pandas.pydata.org/pandas-docs/stable/whatsnew/v1.1.0.html#all-dtypes-can-now-be-converted-to-stringdtype

> If we don't want to make `dtype="string"` breaking, then either we need to come up with a different name for the dtype (not using "string", like "utf8" or "text"), or either we need to delay introducing a default string dtype until after we have agreement on the NA discussions.

Yea none of these options are great...but out of them I still would probably prefer waiting. I think right now we are marching down a path of "string" missing values:

  1. Returning pd.NA today
  2. Returning np.nan with this PDEP (granted those changes are already in main)
  3. Going back to returning pd.NA with the NA PDEP

Reply:
Adding on that this is all done with zero deprecation; using dtype="string" for 4 years has allowed users to intentionally work around the shortcomings of the NumPy type system. As far as I understand this PDEP would mean that:

```python
>>> ser = pd.Series(["foo", "bar", None], dtype="string")
>>> ser.str.isalnum()
0    True
1    True
2    <NA>
dtype: boolean
```

is going to revert back to:

```python
>>> ser.str.isalnum()
0    True
1    True
2    None
dtype: object
```

with no warning or heads up to users who have intentionally been using the "string" data type all this time. 

Even if we agreed as a team that we were OK with going back to the NumPy type system (I don't want to...but speaking hypothetically) not doing any type of deprecation at all just to hit an arbitrary release date is a showstopper for me

Reply:
> is going to revert back to:

My intention is that it would return the following (see https://github.com/pandas-dev/pandas/issues/54805, as one of the design issues listed in https://github.com/pandas-dev/pandas/issues/54792 that we still needed to resolve):
```python
>>> ser.str.isalnum()
0    True
1    True
2    False
dtype: bool
```

which is still different (i.e. not having a missing value in the result), but at least it gives a boolean result (and is consistent with how NaN propagation works in other predicates for our default dtypes). 
Having it as a bool will ensure that typical operations (eg `ser[ser.str.isalnum()]`) will continue to work as expected.

Reply:
While there will definitely be code breakages, I think that pattern will be be the case for many typical workflows using pandas methods. Whether the string column uses NaN or NA, many workflows will continue to work fine because our methods are designed to work with them (eg `isna` will work either way whether the column uses NaN or NA). There are of course specific behaviours that will change, like if you check for the exact scalar for a missing value, or the propagation in comparisons. But unless your code has a specific check for `pd.NA` in it, it is likely going to keep working, I think.

That is not to say that those breakages are not real. But my tradeoff that I am currently making is that I think this is worth it for the benefit of having a "string" dtype in 3.0 for _all_ users at the cost of some breakage for people using the experimental dtype (and note, they will always be able to update their code to continue to use those NA-based dtypes, by slightly changing the way the dtype is specified, but then without having to change anything else in their code).

In the end:

> Yea none of these options are great...but out of them I still would probably prefer waiting.

is of course a completely fine stance, and it's OK to disagree on that trade-off (and my PDEP is proposing to not wait).

Reply:
> Having it as a bool will ensure that typical operations (eg `ser[ser.str.isalnum()]`) will continue to work as expected.

I am confused - this works today if you use `dtype="string"` but this PDEP is going to break that (?).

We introduced dtype="string" 4 years ago intentionally with pd.NA to avoid these nuances of the NumPy type system. For users that have been current with their use of pandas, this part of the PDEP especially is not true:

> Otherwise, the actual string-specific functionality (such as the `.str` accessor
methods) should all keep working as is. By preserving the current missing value
semantics, this proposal is also backwards compatible on this aspect.

since for those users in particular it is taking a step back...

I also disagree with calling the existing dtype experimental - it has been around for 4 years and served a purpose. The fact that it says experimental in the documentation is just an oversight; it's made its way through a few major releases so far without any material changes

Reply:
I'm also cognizant that this PDEP is not exclusively about missing handling, but the unfortunate fact today is that without PDEP-13 and PDEP-XX coming before this then our data types and na markers are bundled together.

With that being the case, if the intent here is to step back from pd.NA -> np.nan but then later return to pd.NA as a default I think it should be explicitly stated.

Reply:
> We introduced dtype="string" 4 years ago intentionally with pd.NA to avoid these nuances of the NumPy type system. For users that have been current with their use of pandas, this part of the PDEP especially is not true:
> 
> > Otherwise, the actual string-specific functionality (such as the `.str` accessor
> > methods) should all keep working as is. By preserving the current missing value
> > semantics, this proposal is also backwards compatible on this aspect.
> 
> For those users in this is taking a step back...

I think that @WillAyd has a valid point.

If people were NOT using `dtype="string"`, but rather had all strings as `dtype="object"`, then this PDEP maintains the behavior of how strings are processed in 2.x, but just changes the dtype.

But if people had opted-in to using `dtype="string"`, then this PDEP will change the behavior.

It is most likely that there are more people who handle strings with `dtype="object"` than `dtype="string"`., so the comment above about "actual string-specific functionality...should all keep working as is" is true for that group.

But we have a problem with the people who use `dtype="string"` today.

Reply:
I would be fine with changing the behavior for `dtype="string"` and `dtype=pd.StringDtype()`, but would _also_ be OK with only changing it for users who don't specify a dtype, i.e. `pd.Series(['a', 'b', None])`, which IIUC is the only part that is actually in main.

Reply:
> > Having it as a bool will ensure that typical operations (eg `ser[ser.str.isalnum()]`) will continue to work as expected.
> 
> I am confused - this works today if you use `dtype="string"` but this PDEP is going to break that (?).

No, my answer to your example snippet was trying to explain how I would ensure this does _not_ break (if we return `bool` column instead of object dtype with True/False/NaN will ensure that filtering keeps working).


> [@jbrockmendel] would _also_ be OK with only changing it for users who don't specify a dtype, i.e. `pd.Series(['a', 'b', None])`, which IIUC is the only part that is actually in main.

Patrick actually already changed `dtype="string"` to mean the future default dtype as well (so not only the case of inference without specified dtype). 
(whether it is already in main or not is not too important I think, if we feel it is better to change this. But personally I think `dtype="string"` meaning something different than the default string dtype you get without specifying the dtype is going to be very confusing ..)

Reply:
> I would be fine with changing the behavior for `dtype="string"` and `dtype=pd.StringDtype()`, but would _also_ be OK with only changing it for users who don't specify a dtype, i.e. `pd.Series(['a', 'b', None])`, which IIUC is the only part that is actually in main.

Just to make sure that I understand what @jbrockmendel is suggesting in the second half of his statement:
- If a user doesn't specify a dtype, but we infer the dtype to be strings, we use the current "pyarrow_numpy" type, which means storage is "pyarrow" and missing values are `np.nan`.  If `pyarrow` is not installed, it uses python strings and missing values are `np.nan`.  Then, for current users who don't specify a dtype, they get the current behavior with respect to missing values.
- If a user DOES specify `dtype="string"` or `pd.StringDtype()`, we use the current behavior, which means we use missing values as `pd.NA()`.  Then storage is "pyarrow" and if `pyarrow` is not installed, it uses python strings.  But either way, missing values are represented by `pd.NA`.

Reply:
> But personally I think `dtype="string"` meaning something different than the default string dtype you get without specifying the dtype is going to be very confusing ..)

I think we have to carefully specify what the user specifies in a `dtype` argument and how that gets interpreted, versus what we return as the `dtype` when they look at `Series.dtype`.  

So we could have a mapping that says

| User specifies `dtype`= | pandas returns `Series.dtype` |
| ------------------------|-------------------------------|
| Unspecified             | `"string[pyarrow_numpy]"` OR `"string[python]"`     |
| `"string"`              | `"string[pyarrow]"`           |
| StringDtype("pyarrow")  | `"string[pyarrow]"`           |
| StringDtype("python")   | `"string[python]"`            |
| StringDtype("pyarrow_numpy") | `"string[pyarrow_numpy]"` |

The first row depends on whether `pyarrow` is installed.
For the second, third and fifth rows, if `pyarrow` is not installed, we raise an Exception.

Separately, we can then debate what the values in the second column should look like in #58613  .  I personally am not a fan of `"pyarrow_numpy"`

Reply:
> No, my answer to your example snippet was trying to explain how I would ensure this does _not_ break (if we return `bool` column instead of object dtype with True/False/NaN will ensure that filtering keeps working).

Ah OK - I didn't realize you were proposing that change be a part of this PDEP, just thought it was an idea you had for the future. But that's a completely new behavior...and then begs the question of do we go back and change dtype=object to have that same behavior or just have dtype="string" exclusively have it. Ultimately we end up with the same issue

Reply:
Yeah, I also agree with Will that it's not fair to change this without warning for people already using "string".
(pd.NA is also a big selling point of the ``dtype="string"`` too)

Maybe a good compromise would be to use ``string[pyarrow]`` under the hood  for those users (if they had it installed)?

If we were to move ahead with the move to nullable dtypes in general, I worry that this changing of the na value for ``dtype="string"`` from pd.NA -> np.nan -> pd.NA will cause a lot of confusion.

If we were to do 2.3 (like I suggested below), this might be addressable there (with a deprecation).

Reply:
Still adding some deprecation warnings in 2.x for current users of StringDtype is something we certainly could do. I am personally ambivalent about it, but fine with adding it if others think that is better (I do think it might become quite noisy, and it also does not change the fact that 3.0 will switch from NA to NaN)

Reply:
> I think we have to carefully specify what the user specifies in a `dtype` argument and how that gets interpreted, versus what we return as the `dtype` when they look at `Series.dtype`.
> 
> So we could have a mapping that says

I created a variant of that table https://github.com/pandas-dev/pandas/issues/58613#issuecomment-2100453677 with a concrete proposal

> For the second, third and fifth rows, if `pyarrow` is not installed, we raise an Exception.

(for clarity, this "second" row referred to specifying a dtype with `"string"`) 
If you explicitly ask for pyarrow, then yes raising an exception is fine and expected. But a generic `"string"` (or `StringDtype()`) has to mean "whatever string dtype that is the default" and so cannot raise an exception if pyarrow is not installed, but should return the object-dtype based fallback.